pkg/controllers/nodeoverlay/controller.go

```diff
@@ -0,0 +1,256 @@
+		For(&v1alpha1.NodeOverlay{}).
+		Watches(&v1.NodePool{}, NodeOverlayEventHandler(c.kubeClient)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+			RateLimiter:             reasonable.RateLimiter(),
+		})
+
+	for _, nodeClass := range c.cloudProvider.GetSupportedNodeClasses() {
+		b.Watches(nodeClass, NodeOverlayEventHandler(c.kubeClient))
+	}
+
+	return b.Complete(c)
+}
+
+func (c *Controller) checkOverlayPerNodePool(nodePool v1.NodePool, its []*cloudprovider.InstanceType, overlay v1alpha1.NodeOverlay) []string {
+	overlaysWithConflict := sets.Set[string]{}
+	overlayRequirements := scheduling.NewNodeSelectorRequirements(overlay.Spec.Requirements...)
+
+	for _, it := range its {
+		offerings := getOverlaidOfferings(nodePool, it, overlayRequirements)
+		// if we are not able to find any offerings for an instance type
+		// This will mean that the overlay does not select on the instance all together
+		if len(offerings) == 0 {
+			continue
+		}
+
+		conflictingPriceOverlay := c.isPriceUpdatesConflicting(nodePool.Name, it.Name, offerings, overlay)
+		if len(conflictingPriceOverlay) == 0 {
+			c.temporaryStore.updateInstanceTypeOffering(nodePool.Name, it.Name, overlay, offerings)
+		}
+		conflictingCapacityOverlay := c.isCapacityUpdatesConflicting(nodePool.Name, it.Name, overlay)
+		if len(conflictingCapacityOverlay) == 0 {
+			c.temporaryStore.updateInstanceTypeCapacity(nodePool.Name, it.Name, overlay)
+		}
+		// When we find an instance type that is matches a set offering, we will track that based on the
+		// overlay that is applied
+		if len(conflictingPriceOverlay) != 0 || len(conflictingCapacityOverlay) != 0 {
+			overlaysWithConflict.Insert(overlay.Name)
+			overlaysWithConflict.Insert(conflictingPriceOverlay...)
+			overlaysWithConflict.Insert(conflictingCapacityOverlay...)
+		}
+	}
+
+	return overlaysWithConflict.UnsortedList()
+}
+
+// getOverlaidOfferings will validate that an instance type matches a set of node overlay requirements
+// if true, the set of Compatible offering. This function effectively assumes, that if the if there are no offering returned then
+// the instance type is not Compatible with the overlay requirements. In cases, were an capacity overlay is being intended to be applied
+// based on the offerings, this will be an all or nothing operation. If one offering matches to the requirements
+// it will be applied at the instance type level or all the offerings.
+func getOverlaidOfferings(nodePool v1.NodePool, it *cloudprovider.InstanceType, overlayReq scheduling.Requirements) cloudprovider.Offerings {
+	// The additional requirements will be added to the instance type during scheduling simulation
+	// Since getting instance types is done on a NodePool level, these requirements were always assumed
+	// to be allowed with these instance types.
+	instanceTypeRequirements := scheduling.NewRequirements(
```

Comment:
One potential optimization we could make here is first checking if the overlay is compatible with the NodePool (and skipping all ITs if not, then checking the specific instance requirements. I'm not too concerned about this, but it's worth considering.

Reply:
I'm not sure this makes much of a difference, since this will just move the check up one level. You will still have to do the same amount of check there